design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,890 @@
+    }
+  }
+}
+
+struct SWManager {
+  sender: Recipient<SoftwareRequest>
+}
+
+impl SWManager {
+  pub async fn do_this(&mut self, arg: ThisArg) -> Result<(), ChannelError> {
+    let req : SoftwareRequest = arg.into();
+    self.sender.send(req).await
+  }
+
+  pub async fn do_that(&mut self, arg: ThatArg) -> Result<(), ChannelError> {
+    let req : SoftwareRequest = arg.into();
+    self.sender.send(req).await
+  }
+}
+
+struct OpScheduler {
+  sender: Recipient<OperationRequest>
+}
+
+impl OpScheduler {
+  pub async fn say(&mut self, arg: &str) -> Result<(), ChannelError> {
+    let req : OperationRequest = arg.into();
+    self.sender.send(req).await
+  }
+}
+```
+
+To be discussed:
+* Can these smart handlers be generated by a macro?
+
+#### Abstracting channels from peers
+
+#### Actor with no inputs or outputs
+
+An actor can have no input messages and only acts as a source of output messages.
+Similarly, an actor can have not output messages and be a sink of input messages.
+Notable examples are respectively a source of measurements and a message logger.
+
+This can be represented using an `enum` with no constructors and hence with no values.
+
+```rust
+/// An actor can have no input or no output messages
+#[derive(Clone, Debug)]
+pub enum NoMessage {}
+
+impl Actor for Collectd {
+  type Input = NoMessage;
+  type Output = Measurement;
+  
+  // ...
+}
+
+impl Actor for Logger {
+  type Input = LogEntry;
+  type Output = NoMessage;
+  
+  // ...
+} 
+```
+
+#### Out-of-band runtime messages
+
+An actor should also be able to handle requests sent by the runtime, typically a shutdown request.
+
+* For observability purposes, all these requests are materialized by messages sent to the actor.
+* These requests can be handled by specific methods of the `Actor` trait, as a `shutdown()` method.
+  However, it might be simpler to let the actor implementations processes these messages as any other messages.
+* These runtime messages must be handled with a higher priority as regular ones.
+  A shutdown request should not wait for all pending messages to be processed first.
+* This priority mechanism must be encapsulated by the actor mailbox,
+  to be able to improve prioritization of message delivery.
+* A simple option to do that is to use two `mpsc` channels for a mailbox,
+  one for high-priority messages, the other for regular messages.
+* Using a [`biased tokio::select!`](https://docs.rs/tokio/latest/tokio/macro.select.html#fairness)
+  ensures that high-priority messages will be delivered first to the actor.
+* Sending high-priority messages could be open to regular actor peers.
+  However, it must then be clear that there is a risk for regular messages to be stalled.
+
+#### Size of the channel buffers
+
+* With unbounded channels, there is a risk of actors being overflowed
+  and failing to catch up till memory exhaustion.
+* Bounded channels enable back-pressure from slow actors onto fast ones,
```

Comment:
The decision of unbounded or bounded channels might depend on the function.

It is important that we also support blocking style communication for specific actors (or messages?) so we can make the actors more composable. For some operations such as firmware, software, configuration, these generally should be processed serially to make the handling of those side-effect heavy artefacts more predictable (and generally the order does matter, so we want this to be repeatable).

Also by supporting serial processing of some messages means that operations can also be "queued" on the cloud side, and the user can be reassured that the operations will be processed in the order that they were created.

Reply:
That's a good point: the decision must be delegated to the code that connects the actors.